### DISCUSSION TITLE
Is it me or react-aria have a larger bundle-size than similar libraries?

### DISCUSSION THREAD

**Discussion by johny-ss** [adobe/react-spectrum#5636] [Q&A]
I left headless UI out of the comparison since it is not intended to be such a complete set of components and the comparison would be unfair.

[react-aria.components](https://bundlephobia.com/package/react-aria-components@1.0.0) is 165 kb.

[Ariakit](https://bundlephobia.com/package/ariakit@2.0.0-next.44) is 39 kb. I know it has fewer components and functionalities, but I don't think it can be explained by that reason alone, since it is 25% the size of RAC.

radix is difficult to compare since they do not have a package with all the components. I tried the following options:
[@radix-ui/themes](https://bundlephobia.com/package/@radix-ui/themes@2.0.3): 80 kb.  I see that it has many components, I don't know if all of them
[radix-ui](https://bundlephobia.com/package/radix-ui@1.0.1): 49 kb. A year ago someone made a package grouping all the radix components

I would have thought that since react-aria was lower level and shared so many hooks between components, the bundle size would be smaller, but it doesn't give me that feeling.

Could it be that react-aria's size is not optimized?

**Comment by devongovett** [comment]
Hard to compare apples to apples when you are comparing the size of the entire library. React Aria has a lot more features than the ones you're comparing it with, so the total size of everything is going to be larger no matter what. You would need to compare the sizes of individual components rather than the whole thing.

**Reply by johny-ss** [reply]
Ok, I made the following comparison of some components with https://bundlejs.com/

I tried to be generous with react-aria. For example, in menu I didn't add the button component even though its counterparts have `MenuButton`, or in ariaKit's Popover I left `popoverArrow, PopoverHeading, and PopoverDescription`.

```js
// Button
export { Button } from "@ariakit/react"; // 7.13
export {Button} from "react-aria-components"; // 64.6

// Menu
export {Menu, MenuBar, MenuButton, MenuItem, MenuProvider, MenuSeparator} from "@ariakit/react"; // 78.5
export * from "@radix-ui/react-dropdown-menu@2.0.6"; // 70
export {Menu, MenuItem, MenuTrigger, Popover} from 'react-aria-components';  // 88.9

// Modal dialog
export {Dialog, DialogProvider, DialogDisclosure,DialogDismiss, DialogHeading, DialogDescription} from "@ariakit/react"; // 59.7
export * from "@radix-ui/react-dialog@1.0.5"; // 55.7
export {Dialog, DialogTrigger, Modal} from 'react-aria-components'; // 68.6

// Popover
export {PopoverProvider, PopoverAnchor, PopoverDisclosure, Popover, PopoverArrow, PopoverHeading, PopoverDescription, PopoverDismiss} from "@ariakit/react"; // 68.4
export * from "@radix-ui/react-popover@1.0.7"; // 65.1
export {Popover} from 'react-aria-components'; // 70.2

// Tooltip
export {TooltipProvider, TooltipAnchor, Tooltip, TooltipArrow} from "@ariakit/react"; // 69.3
export * from "@radix-ui/react-tooltip@1.0.7"; // 60.5
export {Tooltip, TooltipTrigger} from 'react-aria-components'; // 66
```

except in `Tooltip`, RAC usually has ~10-20kb more per component, and in `Button` about 57kb more.

It occurs to me that a possible reason for this may be largely due to the effort to get RAC to work in browsers that do not support `pointerEvents`, which by the way seems to already have [enough support between browsers](https://caniuse.com/pointer).

What do you think?

**Reply by reidbarber** [reply]
Providing a fallback for pointer events is just one feature of `Button`. I recommend reading the [Building a Button](https://react-spectrum.adobe.com/blog/building-a-button-part-1.html) blog post series to see a more thorough explanation of the features provided. We definitely aim to minimize the bundle size as much as possible, but we put an emphasis on accessibility, UX, and DX. If there are avenues to reduce bundle size without sacrificing those, we are definitely interested.

**Reply by johny-ss** [reply]
> If there are avenues to reduce bundle size without sacrificing those, we are definitely interested.

Well, for example I see that pointer events [are already supported](https://caniuse.com/pointer) in all the browsers that [RA intends to support](https://react-spectrum.adobe.com/react-aria/interactions.html#supported-browsers-and-platforms), so eliminating said fallbacks could be one way.

On the other hand, do you know of any flaws in accessibility, UX or DX in the Radix dropdown-menu? It is 19 kb less than the RAC one.

**Reply by reidbarber** [reply]
We have an existing issue for doing some cleanup in `usePress` at https://github.com/adobe/react-spectrum/issues/1720 (includes removing the fallback for pointer events).

---

I think all the libraries you mentioned are great, and are preferable to building these features from scratch. I think doing a detailed comparison of the menu components would require some time to test across an array of devices and assistive technology. In terms of DX, I would highlight React Aria's [Styling](https://react-spectrum.adobe.com/react-aria/styling.html) and [Advanced Customization](https://react-spectrum.adobe.com/react-aria/advanced.html) sections of the docs. Also, features such as [internationalization](https://react-spectrum.adobe.com/react-aria/internationalization.html) and [drag and drop](https://react-spectrum.adobe.com/react-aria/dnd.html) may be important to your apps, so consider those features as well.

**Reply by devongovett** [reply]
fwiw, there are some tree-shaking issues at play here as well. I believe bundlejs uses esbuild under the hood. I submitted an issue over there for one issue that affected bundle size a lot: https://github.com/evanw/esbuild/issues/3561. This was fixed in the latest version but not sure if it has made its way to bundlejs yet. Another issue was #5645 which caused textfield to always be included no matter if it was used or not, which should be fixed in our next release. Those two changes seemed to make it a lot smaller, but I'm sure there's more we can do.

**Reply by devongovett** [reply]
In regards to pointer events, I think we could remove the fallbacks, but my guess is that a lot of unit tests rely on them still since JSDOM still does not support them (https://github.com/jsdom/jsdom/issues/2527). But perhaps we could gate the fallback behind a `process.env.NODE_ENV === "test"` check or something so it gets removed in production builds.

**Reply by johny-ss** [reply]
Thanks to both of you. I hope those changes on the way can reduce the bundle size 🤞

I love that RA does so much for web accessibility. However, an extra 10-20kb per component or 64kb for a button is definitely something I'm worried about.

> In regards to pointer events, I think we could remove the fallbacks, but my guess is that a lot of unit tests rely on them still since JSDOM still does not support them (https://github.com/jsdom/jsdom/issues/2527). But perhaps we could gate the fallback behind a process.env.NODE_ENV === "test" check or something so it gets removed in production builds.

Yes, that or move those tests to a real browser. Maybe they can be parallelized by mounting and unmounting on the same page to make it faster.

**Reply by devongovett** [reply]
Just tested, and after the changes, button is 8.1KB minified with esbuild and brotlied. But also keep in mind that the components share a lot of dependencies with each other (e.g. usePress), so if you use more than one component it will amortize a lot. You can't calculate the total size by adding up the individual components, you'd need to bundle the exact combination you are using.

> Yes, that or move those tests to a real browser. 

Would be nice, but I'm not referring to our tests, but all of the applications using React Aria. We can't control if they use JSDOM or not, and I think a majority do (it's the default in Jest for example).

**Comment by axelboc** [comment]
I've just tried implementing a very basic menu component with `react-aria-components@1.2.0`. I'm super impressed with the composable API, the detailed documentation and the UX capabilities for end users. :heart_eyes: So, thanks so much for that.

The only surprise is that the code below basically adds **185 kB** to my Vite bundle, or **50 kB** gzipped:

```jsx
import {
  Button,
  Menu,
  MenuItem,
  MenuTrigger,
  Popover,
} from 'react-aria-components';

<MenuTrigger>
  <Button>Foo</Button>
  <Popover>
    <Menu>
      <MenuItem>Bar</MenuItem>
    </Menu>
  </Popover>
</MenuTrigger>
```

When I analyze the bundle, I see a lot of stuff that aren't used, like `@react-aria/i18n`, `@react-aria/selection`, `@react-stately/selection`, `@react-stately/tree`, `useSubmenuTrigger` in `@react-aria/menu`, etc.

I don't deny that the library is powerful, especially compared to the one I'm trying to replace ([`react-aria-menubutton`](https://github.com/davidtheclark/react-aria-menubutton)) but this is quite extreme. I'd really like to avoid adding to the [JS bloat problem](https://tonsky.me/blog/js-bloat/) if I can. I'll keep an eye on future releases in case tree-shaking issues are resolved in the future. :pray: 

> For what it's worth, I tried Ariakit as well and found an increase of 128 kB (38 kB gzipped), which isn't much better.

**Reply by devongovett** [reply]
Those _are_ all used, they are dependencies of those components. Submenu might be able to be tree shaken when not used, but not sure how big of a difference it would make. Perhaps there are more ways we can improve the size, but good UX unfortunately requires a lot of code (I wouldn't call that "bloat") – at least until the browser gives us more tools built in (popover is coming).

**Reply by reidbarber** [reply]
It has been discussed before, but I want to emphasize the property of having so much shared code between components, which doesn't show itself when testing one component. For instance `@react-aria/selection` and `@react-stately/selection` get used by all components that feature selection, so you'll only load those once no matter how many components with selection you're using.  Same for things like press interactions, focus handling, and more. The "Law of Diminishing Bundle Size Increases" I guess.
